sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala

```diff
@@ -424,169 +395,101 @@ class GpuCoalesceIterator(iter: Iterator[ColumnarBatch],
-  private def decompressBatches(): Unit = {
-    if (compressedBatchIndices.nonEmpty) {
-      val compressedVecs = compressedBatchIndices.map { batchIndex =>
-        batches(batchIndex).column(0).asInstanceOf[GpuCompressedColumnVector]
-      }
-      if (codec == null) {
-        val descr = compressedVecs.head.getTableMeta.bufferMeta.codecBufferDescrs(0)
-        codec = TableCompressionCodec.getCodec(descr.codec)
-      }
-      withResource(codec.createBatchDecompressor(maxDecompressBatchMemory)) { decompressor =>
-        compressedVecs.foreach { cv =>
-          val bufferMeta = cv.getTableMeta.bufferMeta
-          // don't currently support switching codecs when partitioning
-          val buffer = cv.getBuffer.slice(0, cv.getBuffer.getLength)
-          decompressor.addBufferToDecompress(buffer, bufferMeta)
+  private[this] def popAllDecompressed(): Array[ColumnarBatch] = {
+    val wip = batches.map(_.getColumnarBatch())
+    batches.safeClose()
+    batches.clear()
+
+    try {
+      val compressedBatchIndices = wip.zipWithIndex.filter { pair =>
+        GpuCompressedColumnVector.isBatchCompressed(pair._1)
+      }.map(_._2)
+      if (compressedBatchIndices.nonEmpty) {
+        val compressedVecs = compressedBatchIndices.map { batchIndex =>
+          wip(batchIndex).column(0).asInstanceOf[GpuCompressedColumnVector]
+        }
+        if (codec == null) {
+          val descr = compressedVecs.head.getTableMeta.bufferMeta.codecBufferDescrs(0)
+          codec = TableCompressionCodec.getCodec(descr.codec)
         }
-        withResource(decompressor.finish()) { outputBuffers =>
-          outputBuffers.zipWithIndex.foreach { case (outputBuffer, outputIndex) =>
-            val cv = compressedVecs(outputIndex)
-            val batchIndex = compressedBatchIndices(outputIndex)
-            val compressedBatch = batches(batchIndex)
-            batches(batchIndex) = MetaUtils.getBatchFromMeta(outputBuffer, cv.getTableMeta)
-            compressedBatch.close()
+        withResource(codec.createBatchDecompressor(maxDecompressBatchMemory)) { decompressor =>
+          compressedVecs.foreach { cv =>
+            val bufferMeta = cv.getTableMeta.bufferMeta
+            // don't currently support switching codecs when partitioning
+            val buffer = cv.getBuffer.slice(0, cv.getBuffer.getLength)
+            decompressor.addBufferToDecompress(buffer, bufferMeta)
+          }
+          withResource(decompressor.finish()) { outputBuffers =>
+            outputBuffers.zipWithIndex.foreach { case (outputBuffer, outputIndex) =>
+              val cv = compressedVecs(outputIndex)
+              val batchIndex = compressedBatchIndices(outputIndex)
+              val compressedBatch = wip(batchIndex)
+              wip(batchIndex) = MetaUtils.getBatchFromMeta(outputBuffer, cv.getTableMeta)
+              compressedBatch.close()
+            }
           }
         }
       }
+    } catch {
+      case t: Throwable =>
+        wip.safeClose()
```

Comment:
Suggest using `closeOnExcept` to make this a bit cleaner.

Reply:
I am trying to remember, either `closeOnExcept` didn't work with `safeClose` or it turned the array into a sequence and cause some other code issues.  I can look at it again if needed.

Reply:
`wip` should be another `ArrayBuffer` which `closeOnExcept` explicitly handles, so I'm not sure why it wouldn't work.  This isn't mustfix, but it would be good to understand why it doesn't work and potentially fix `closeOnExcept` in a followup PR if there's something it's not doing properly.